specs/anon-aadhaar/specs.md

```diff
@@ -0,0 +1,542 @@
+2. Selective Disclosure, verify an attributes from the four revealed fields.
+
+Output format:
+
+```
+{
+  proof: Bytes,            // Zero-knowledge proof
+  pubKeyHash: Bytes32,     // RSA key commitment
+  timestamp: Uint64,       // UTC Unix timestamp
+  nullifierSeed: string    // Seed of the nullifier
+  nullifier: Bytes32,      // Unique identifier
+  signalHash: Bytes32,     // Optional binding
+  ageAbove18: Boolean       // Optional disclosures
+  gender: string           // Optional disclosures
+  pincode: string          // Optional disclosures
+  state: string            // Optional disclosures
+}
+```
+
+## Error Handling
+
+Implementations MUST handle:
+
+1. Invalid QR format
+2. Invalid signature
+3. Duplicate nullifier
+4. Invalid proof
+
+Error responses MUST include:
+
+1. Error code
+2. Error message
+3. Error details (when available)
+
+# Security Considerations
+
+## Threat Model
+
+The protocol assumes:
+
+1. Trusted setup for zero-knowledge system
+2. Secure and honest UIDAI RSA key, the trust assumption is that the UIDAI RSA private key will not sign non legitimate identities
+3. Secure random number generator
+4. Honest verifier
+
+Known limitations:
+
+1. QR code format changes, if the UIDAI change the format of the Aadhaar Secure QR code, then the protocol must be updated to follow the new data serialization.
+2. Circuit size constraints, the current reference implementation in Circom have around 1 million constraints
+3. Nullifier rotation, the nullifier is based on the Aadhaar photo of the user, this photo can be changed following a process specified by the UIDAI. From the documentation, the photo can be changed within 7 days, meaning that a user could potentially generates 2 nullifiers using the same nullifier Seed. This must be took into consideration when building use cases that requires a nullifier lasting for more than 7 days.
+
+## Privacy Guarantees
+
+The protocol MUST guarantee:
+
+1. Base identity privacy
+2. Nullifier unlinkability, from all the actors except the issuer of the signature
+3. Signal binding when requested
+
+# Implementation Notes
+
+Basic proof generation:
+
+```typescript
+const args = await generateArgs({
+  qrData: QRData,
+  certificateFile: certificate, // x509 PEM certificate of the signing public key
+  nullifierSeed: number,
+});
+
+const anonAadhaarProof = await prove(args);
+```
+
+Basic verification:
+
+```typescript
+const verified = await verify(anonAadhaarProof);
+```
+
+# References
+
+1. [UIDAI QR Code Specification](https://uidai.gov.in/images/resource/User_manulal_QR_Code_15032019.pdf)
+2. RSA PKCS#1 v2.1
+3. SHA-256 FIPS 180-4
+
+# Appendix A: Circuit Architecture
+
+## Components
+
+### 1. AadhaarQRVerifier
+
+**Description**: Verifies the Aadhaar QR data using RSA signature.
```

Comment:
I would maybe say that further info can be found in the appendix.

Reply:
Yes! 
The appendix are referenced above

Reply:
Woops, sorry I think I missed my line count!